src/mc-text-search-str-encode.c

```diff
@@ -0,0 +1,304 @@
+    return generate_prefix_or_suffix_tree(base_str, unfolded_codepoint_len, spec->lb, spec->ub, false);
+}
+
+static mc_substring_set_t *generate_prefix_tree(const mc_utf8_string_with_bad_char_t *base_str,
+                                                uint32_t unfolded_codepoint_len,
+                                                const mc_FLE2PrefixInsertSpec_t *spec) {
+    return generate_prefix_or_suffix_tree(base_str, unfolded_codepoint_len, spec->lb, spec->ub, true);
+}
+
+static uint32_t calc_number_of_substrings(uint32_t strlen, uint32_t lb, uint32_t ub) {
+    // There are len - i + 1 substrings of length i in a length len string.
+    // Therefore, the total number of substrings with length between lb and ub
+    // is the sum of the integers inclusive between A = len - ub + 1 and B = len - lb + 1,
+    // A <= B. This has a closed form: (A + B)(B - A + 1)/2.
+    if (lb > strlen) {
+        return 0;
+    }
+    uint32_t largest_substr = MIN(strlen, ub);
+    uint32_t largest_substr_count = strlen - largest_substr + 1;
+    uint32_t smallest_substr_count = strlen - lb + 1;
+    return (largest_substr_count + smallest_substr_count) * (smallest_substr_count - largest_substr_count + 1) / 2;
+}
+
+static mc_substring_set_t *generate_substring_tree(const mc_utf8_string_with_bad_char_t *base_str,
+                                                   uint32_t unfolded_codepoint_len,
+                                                   const mc_FLE2SubstringInsertSpec_t *spec) {
+    // 16 * ceil(unfolded len / 16)
+    uint32_t cbclen = 16 * (uint32_t)((unfolded_codepoint_len + 15) / 16);
+    if (unfolded_codepoint_len > spec->mlen || cbclen < spec->lb) {
+        // No valid substrings, return empty tree
+        return NULL;
+    }
+    uint32_t folded_codepoint_len = base_str->codepoint_len - 1;
+    // If mlen < cbclen, we only need to pad to mlen
+    uint32_t padded_len = MIN(spec->mlen, cbclen);
+    // Total number of substrings -- i.e. the number of valid substrings IF the string spanned the full padded length
+    uint32_t msize = calc_number_of_substrings(padded_len, spec->lb, spec->ub);
+    uint32_t n_real_substrings = calc_number_of_substrings(folded_codepoint_len, spec->lb, spec->ub);
+    // If real_substrings and msize differ, we need to insert padding, so allocate one extra slot.
+    mc_substring_set_t *set =
+        mc_substring_set_new(base_str, n_real_substrings == msize ? n_real_substrings : n_real_substrings + 1);
+    uint32_t idx = 0;
+    // If folded len < LB, there are no real substrings, so we can skip (avoiding underflow via folded len - LB)
+    if (folded_codepoint_len >= spec->lb) {
+        for (uint32_t i = 0; i < folded_codepoint_len - spec->lb + 1; i++) {
+            for (uint32_t j = i + spec->lb; j < MIN(folded_codepoint_len, i + spec->ub) + 1; j++) {
+                mc_substring_set_insert(set, i, j, idx++, 1);
```

Comment:
I think we need to remove duplicate substrings from this set, so we'll need some kind of unordered set/prefix tree data structure here.

Reply:
Oh, so that we don't return duplicate results when we search for a particular substring... Makes sense. I'll use a trie instead to dedupe.